doc/dev/radosgw/virtual_hosting.md

```diff
@@ -0,0 +1,122 @@
+# radosgw virtual hosting
+
+allow the different rest apis to use their own domain names. route requests between them based on Host header
+
+## motivation
+
+radosgw currently serves all rest apis from the same namespace. some are hosted at the root, with others under paths like `admin`, `auth`, `swift` etc. most apis are enabled by default, so parts of the root s3 namespace are unavailable because those paths get routed to other apis
+
+s3 shares the root path with several other apis (iam, s3control, sts, sns) and relies on tricks to disambiguate the requests and route them correctly. these tricks are based on assumptions that may not be reliable, especially as the apis evolve
+
+to avoid these conflicts in path-based routing, we can first route requests by hostname
+
+this _could_ be accomplished by deploying separate radosgw processes for each api and configuring dns accordingly. however,
+* radosgw processes are heavy-weight, having several background threads and a main pool of 512 threads
+* the overall load would not be well-balanced over radosgws, with some apis expecting very little traffic
+* the metadata cache broadcasts invalidation messages, so will limit scaling to many radosgws
+
+radosgw can continue serving all of these apis from the same frontend socket by instead routing the requests according to their http `Host` header. this [virtual hosting](https://en.wikipedia.org/wiki/Virtual_hosting) technique is common in web servers
+
+## existing request routing
+
+config options:
+
+```
+rgw_enable_apis = s3, s3control, s3website, swift, swift_auth, admin, sts, iam, notifications
+rgw_dns_name =
+rgw_dns_s3website_name =
+```
+
+request flow:
+
+1. `process_request()` calls `get_manager()` on the root `RGWRESTMgr` to route the given request path to its final `RGWRESTMgr`
```

Comment:
No suggestion is popping into my mind at the moment, but maybe the name could/should more closely reflect that it's doing virtual host-based routing? I sorta see that it's part of REST but when I think of REST I think more of managing state and whatnot, rather than URL- or host-name stuff.